File: algorithms/dp/maximum_product_subarray.py
# https://leetcode.com/problems/maximum-product-subarray/
# https://leetcode.com/problems/maximum-product-subarray/discuss/183483/JavaC%2B%2BPython-it-can-be-more-simple

# time complexity: O(n)

class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        rev_nums = nums[::-1]
        for i in range(1, len(nums)):
            nums[i] *= nums[i - 1] or 1
            rev_nums[i] *= rev_nums[i - 1] or 1
        return max(nums + rev_nums)

File: leetcode/dp/maximum_product_subarray.cc
// https://leetcode.com/problems/maximum-product-subarray/
// https://leetcode.com/problems/maximum-product-subarray/discuss/183483/JavaC%2B%2BPython-it-can-be-more-simple

// time complexity: O(n)

#include <vector>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int> &nums) {
        int len = nums.size(), max_p = INT32_MIN, l = 0, r = 0;
        for (int i = 0; i < len; i++) {
            l = (l ? l : 1) * nums[i];
            r = (r ? r : 1) * nums[len - 1 - i];
            max_p = max(max_p, max(l, r));
        }
        return max_p;
    }
};